File: S2_KP/Lectures/codes02/Abstrakcyjny.cpp
#include<iostream>

using namespace std;

// Klasa jest abstrakcyjna jesli ma metode abstrakcyjna
class Pojazd {    
public:
	// Sposob definiowania metody abstrakcyjnej
	virtual void PodajCene()=0;
};

class Ciezarowy : public Pojazd{
public:
	int PodajLadownosc(){
     return 500;
    }   
	 void PodajCene(){
        cout << "30000" << endl;
    }
};

class SamochodDostawczy : public Pojazd{   
public:
	int PodajIloscMiejsc(){
        return 8;
    }
    int PodajLadownosc(){
        return 500;
    }
    void PodajCene(){
        cout << "60000" << endl;
    }
};

int main(int argc, char* argv[]) {

	Ciezarowy *c = new Ciezarowy();
	Pojazd *p = new SamochodDostawczy();
	p->PodajCene();

	// Niepoprawnie. Nie mozna tworzyc obiektu dla klasy abstrakcyjnej
	//Pojazd *p2 = new Pojazd();
	
	// Moze byc tablica dla typow abstrakcyjnych
	Pojazd* p1[2] = {c,p};
	p1[0]->PodajCene();

	Pojazd** p2 = new Pojazd*[3];
	p2[0]=p;
	p2[1]=c;

	p1[0]->PodajCene();
	p1[1]->PodajCene();

}

File: S2_KP/Lectures/codes02/Abstrakcyjny.java
// Klasa abstrakcyjna. Nie mozna tworzyc obiektu dla niej.
abstract class Pojazd {    
    public int Cena;
    // Metoda abstrakcyjna, bez implementacji
    public abstract int PodajCene();
    // Metoda zwykla z implementacja;
    public void PodajCene1(int cena){
        this.Cena=cena;
    };

}
class Ciezarowy extends Pojazd{
    public int PodajLadownosc(){
    	return 500;};
    
    public int PodajCene() {
    return 500;}
}
class SamochodDostawczy extends Pojazd{   
    public int PodajIloscMiejsc() {
    return 8;}
    
    public int PodajLadownosc() {
    return 500; }
    
    public int PodajCene(){
       return 60000;}
}
public class Abstrakcyjny {
    public static void main(String[] args) {

         // Nie mozna stworzyc obiektu, bo Pojazd to klasa abstrakcyjna
        //Pojazd s2 = new Pojazd();
         
        // Tworzenie obiektow od klas nieabstrakcyjnych
        Pojazd s = new Ciezarowy();
	    s.PodajCene();	
        Pojazd s1 = new SamochodDostawczy();
        s.PodajCene1(100);

       
        // Mozna stworzyc tablice dla Pojaz i przypisac obiekty tego typu
        Pojazd p[] = new Pojazd[2];     
        p[0]=s;
        p[1]=s1;
            
        System.out.println(p[0].PodajCene());
        System.out.println(p[1].PodajCene());
    }
}